In [3]:
import cProfile

In [4]:
import main

INFO:tensorflow:Enabling eager execution
INFO:tensorflow:Enabling v2 tensorshape
INFO:tensorflow:Enabling resource variables
INFO:tensorflow:Enabling tensor equality
INFO:tensorflow:Enabling control flow v2


In [5]:
player = main.MCTSPlayer3(num_evals= 100000, curiosity = 0.001, name="saved_model/2021-08-23-v3-model",)

In [7]:
main.run_game(players=[player, player],truncate_moves=2)

          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++++●+++++++++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 0
Ball at: [7 9]

G11
G9
F13
H11
H12
NW 
G9
E7
SE NE E 
G13
F15
C5
A3
F8 0.43852758407592773 1
F9 0.44821083545684814 1
F10 0.45043623447418213 1
F11 0.4411371946334839 1
F12 0.4481600522994995 1
G8 0.43863242864608765 1
G9 0.3974459171295166 1
G10 0.43550336360931396 1
G11 0.5835057814654551 1672
G12 0.4473913311958313 1
H8 0.4276010990142822 1
H9 0.3729366064071655 1
H11 0.5217154496594479 19
H12 0.4511333703994751 1
I8 0.4419248104095459 1
I9 0.36610448360443115 1
I10 0.4500507116317749 1
I11 0.4764941334724426 1
I12 0.4547402858734131 1
J8 0.4379305839538574 1
J9 0.43893271684646606 1
J10 0.4477374553

In [10]:
cProfile.run('main.run_game(players=[player, player],truncate_moves=2)', '2021-09-16-stats')

          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++++●+++++++++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 0
Ball at: [7 9]

G11
G9
F13
H11
H12
NW 
G9
E7
SE NE E 
G13
F15
C5
A3
F8 0.43852758407592773 1
F9 0.44821083545684814 1
F10 0.45043623447418213 1
F11 0.4411371946334839 1
F12 0.4481600522994995 1
G8 0.43863242864608765 1
G9 0.3974459171295166 1
G10 0.43550336360931396 1
G11 0.5835057814654551 1672
G12 0.4473913311958313 1
H8 0.4276010990142822 1
H9 0.3729366064071655 1
H11 0.5217154496594479 19
H12 0.4511333703994751 1
I8 0.4419248104095459 1
I9 0.36610448360443115 1
I10 0.4500507116317749 1
I11 0.4764941334724426 1
I12 0.4547402858734131 1
J8 0.4379305839538574 1
J9 0.43893271684646606 1
J10 0.4477374553

In [8]:
import tensorflow as tf
if tf.test.gpu_device_name():
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))
else:
    print("Please install GPU version of TF")

Please install GPU version of TF


In [11]:
import pstats
from pstats import SortKey
p = pstats.Stats('2021-09-16-stats')

In [12]:
p.sort_stats(SortKey.CUMULATIVE).print_stats(10)

Thu Sep 16 17:22:19 2021    2021-09-16-stats

         47462855 function calls (46532612 primitive calls) in 117.723 seconds

   Ordered by: cumulative time
   List reduced from 268 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000  117.723  117.723 {built-in method builtins.exec}
        1    0.000    0.000  117.723  117.723 <string>:1(<module>)
        1    0.000    0.000  117.722  117.722 /Users/theodore/phutball/main.py:394(run_game)
        2    0.000    0.000  117.713   58.857 /Users/theodore/phutball/main.py:949(make_move)
        2    2.560    1.280  117.707   58.854 /Users/theodore/phutball/main.py:1128(build_parallel_max)
1074593/260805    1.411    0.000   47.941    0.000 /usr/local/anaconda3/envs/sage/lib/python3.9/site-packages/tensorflow/python/util/dispatch.py:203(wrapper)
20244/3374    0.963    0.000   35.758    0.011 /usr/local/anaconda3/envs/sage/lib/python3.9/site-packages/tensorflow/pyth

In [13]:
p.sort_stats(SortKey.TIME).print_stats(10)

Thu Sep 16 17:22:19 2021    2021-09-16-stats

         47462855 function calls (46532612 primitive calls) in 117.723 seconds

   Ordered by: internal time
   List reduced from 268 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   247309   29.554    0.000   29.554    0.000 {built-in method tensorflow.python._pywrap_tfe.TFE_Py_FastPathExecute}
  3741574   13.853    0.000   14.155    0.000 /Users/theodore/phutball/main.py:1337(magic_formula)
   200073   13.008    0.000   14.031    0.000 /Users/theodore/phutball/main.py:129(get_flipped_3d_array)
   618840   12.051    0.000   12.051    0.000 /Users/theodore/phutball/main.py:61(<listcomp>)
   606967    4.300    0.000    5.252    0.000 /usr/local/anaconda3/envs/sage/lib/python3.9/site-packages/tensorflow/python/framework/constant_op.py:70(convert_to_eager_tensor)
   200073    2.695    0.000   30.084    0.000 /usr/local/anaconda3/envs/sage/lib/python3.9/site-packages/tensorflow/python/ops

Total: 117 seconds
    
We spend 13 secs in magic_formula, because we call it 3741574 times

In [14]:
player2 = main.MCTSPlayer3(num_evals= 10000, curiosity = 0.001, name="saved_model/2021-08-23-v3-model",)

In [15]:
main.run_game(players=[player2, player2])

          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++++●+++++++++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N+++++++++++++++++++
O+++++++++++++++++++
Side to move: Left
Moves made: 0
Ball at: [7 9]

G11
H9
G13
W 
H9
I9
E NE SE 
F12
D10
F8 0.43852758407592773 1
F9 0.44821083545684814 1
F10 0.45043623447418213 1
F11 0.4411371946334839 1
F12 0.4481600522994995 1
G8 0.43863242864608765 1
G9 0.3974459171295166 1
G10 0.43550336360931396 1
G11 0.5929694323889954 177
G12 0.4473913311958313 1
H8 0.4276010990142822 1
H9 0.3729366064071655 1
H11 0.5217154496594479 19
H12 0.4511333703994751 1
I8 0.4419248104095459 1
I9 0.36610448360443115 1
I10 0.4500507116317749 1
I11 0.4764941334724426 1
I12 0.4547402858734131 1
J8 0.4379305839538574 1
J9 0.43893271684646606 1
J10 0.447737455368042 1
J11 0.44

H9
H10
H11
E E NW SW 
H11
H9
I13
F7
D5
E9 0.3873962163925171 1
E10 0.39225924015045166 1
E11 0.39543330669403076 1
E12 0.38252848386764526 1
E13 0.3878939151763916 1
F9 0.3852677345275879 1
F10 0.38896816968917847 1
F11 0.3936891555786133 1
F12 0.3946274518966675 1
F13 0.38196539878845215 1
G9 0.50284144282341 1
G10 0.38598960638046265 1
G12 0.3916670083999634 1
H9 0.6813136552061353 154
H10 0.380800724029541 1
H11 0.40126919746398926 1
H12 0.37903547286987305 1
I9 0.4516664743423462 1
I10 0.39748287200927734 1
I11 0.39808428287506104 1
I12 0.38878607749938965 1
I13 0.3847329020500183 1
E14 0.38689959049224854 1
E15 0.38773107528686523 1
F14 0.3852968215942383 1
F15 0.39132314920425415 1
G14 0.38240063190460205 1
G15 0.37447071075439453 1
H14 0.3920280337333679 1
H15 0.38652968406677246 1
I14 0.3837851285934448 1
F6 0.3688686490058899 1
F7 0.384565532207489 1
F8 0.38970816135406494 1
G6 0.37803274393081665 1
G7 0.2739884853363037 1
G8 0.390746533870697 1
H6 0.37201565504074097 1
H7 0.2

H11
G9
H13
NW 
G9
F7
SE E E SE 
D5
B3
F8 0.4398101568222046 1
F9 0.449493408203125 1
F10 0.45155471563339233 1
F11 0.4422556161880493 1
F12 0.44943684339523315 1
G8 0.4399150609970093 1
G9 0.39873433113098145 1
G10 0.43662184476852417 1
G11 0.490412175655365 1
G12 0.44868868589401245 1
H8 0.4288893938064575 1
H9 0.37403857707977295 1
H11 0.6444577863883405 168
H12 0.45241332054138184 1
I8 0.4432075023651123 1
I9 0.3672063946723938 1
I10 0.45116913318634033 1
I11 0.47761446237564087 1
I12 0.4560418128967285 1
J8 0.4392050504684448 1
J9 0.4402153491973877 1
J10 0.44902002811431885 1
J11 0.4458332061767578 1
J12 0.45124173164367676 1
G13 0.4618973135948181 1
G14 0.44804829359054565 1
G15 0.4437403082847595 1
G16 0.4438037872314453 1
G17 0.4433649778366089 1
H13 0.4655696153640747 1
H14 0.44496142864227295 1
H15 0.44569218158721924 1
H16 0.4454925060272217 1
H17 0.4443366527557373 1
I13 0.4680200219154358 1
I14 0.44612908363342285 1
I16 0.44663095474243164 1
I17 0.44313979148864746 1
J13 0

H10
SE 
I10
N 
F10
G12
D8
B6
D6 0.3701406915982564 3
D7 0.36460230747858685 3
D8 0.36581047375996906 3
D9 0.3644341031710307 3
D10 0.35752954085667926 3
E6 0.3656507730484009 3
E7 0.42935625712076825 3
E8 0.38547974824905396 3
E9 0.3609439730644226 3
E10 0.37458030382792157 3
F6 0.38437267144521076 3
F7 0.4352028767267863 3
F9 0.3815767765045166 3
F10 0.37072883049647015 3
G6 0.36776745319366455 3
G7 0.4241303503513336 3
G8 0.36640987793604535 3
G10 0.3688831130663554 3
H6 0.3730366826057434 3
H7 0.3659573992093404 3
H8 0.37122438351313275 3
H9 0.3771878480911255 3
H10 0.4832750800110045 42
E11 0.352522333463033 3
F11 0.4142070412635803 1
G11 0.3923785090446472 1
I7 0.3714299003283183 3
I8 0.3609204888343811 3
I9 0.36353057622909546 3
I10 0.3625976045926412 3
I11 0.36474581559499103 3
F12 0.35884735981623334 3
F13 0.36590025822321576 3
G12 0.3621191581090291 3
G13 0.3652661442756653 3
H12 0.35988837480545044 3
I12 0.36581387122472125 3
I13 0.3973510464032491 3
J9 0.36822205781936646 3


KeyboardInterrupt: 